### ISSUE TITLE
Cloudflare Tunnel (Argo tunnel) as system service in OMV-extras

### ISSUE THREAD

**Issue by henkisdabro** [openmediavault/openmediavault#1034]
Cloudflare provides the [cloudflared argo tunnel client](https://github.com/cloudflare/cloudflared) to route outside traffic to selfhosted applications via tunnels, a very useful and well implemented service, especially when using Cloudflare for Teams with Zero-trust (free for up to 50 users). 

I currently run cloudflared as a docker container, but configuring the settings files, obtaining the initial certificates could be made available perhaps from the OMV interface. It's also likely the fact that it sits in a container, I was thinking it would be possible to install as a system-wide service instead, suitably placed in the OMV-extras section, next to Docker and Yacht etc.

**Comment by fdelucchijr** [comment]
This will be awesome!!! 
Please share your cloudflared docker configuration 🙏🏻.

**Comment by henkisdabro** [comment]
Sure here you go, I've made sure to add all available env variables as comments for reference. Also note that you can't start with the config file and just spin up, you first need to run the initialization commands to authenticate with Cloudflare, so it can download the certificates into your docker container, so ensure to have a bind volume in place first, so you can use it subsequently. Note I have two containers configured in this compose, to be able to spin up another cloudflared-service with the latest version, when running, stop the old version container and experience zero downtime. Hope it helps you.

```
# Check https://hub.docker.com/r/cloudflare/cloudflared for image updates
# Add new internal services (and map to a desired hostname) into the config.yml file for inclusion in the tunnel 
# Follow ingress rules guidelines on: https://developers.cloudflare.com/cloudflare-one/connections/connect-apps/configuration/ingress
# Ensure every docker container (and this one) is on your own created docker network for http://containername:port mapping to work
# Run the dns command below once to create the necessary dns record on Cloudflare (adds a CNAME pointing to the tunnel). You can also create this one manually on Cloudflare if you wish.
# Start/enable the tunnel with the run command and keep on

#Environment VARS
#- TUNNEL_ORIGIN_CERT= #Path to the certificate generated for your origin when you run cloudflared login.
     #- NO_AUTOUPDATE= #Disable periodic check for updates, restarting the server with the new version. (default: false)
     #- TUNNEL_METRICS= #Listen address for metrics reporting. (default: "localhost:")
     #- TUNNEL_METRICS_UPDATE_FREQ=5s #Frequency to update tunnel metrics (default: 5s)
     #- TUNNEL_PIDFILE= #Write the application's PID to this file after first successful connection.
     #- TUNNEL_URL= #Connect to the local webserver at URL. (default: "http://localhost:8080")
     #- TUNNEL_HTTP_HOST_HEADER= #Sets the HTTP Host header for the local webserver. 
     #- TUNNEL_ORIGIN_SERVER_NAME= #Hostname on the origin server certificate.
     #- TUNNEL_ORIGIN_CA_POOL= #Path to the CA for the certificate of your origin. This option should be used only if your certificate is not signed by Cloudflare.
     #- NO_TLS_VERIFY= #Disables TLS verification of the certificate presented by your origin. Will allow any certificate from the origin to be accepted. Note: The connection from your machine to Cloudflare's Edge is still encrypted. (default: false)
     #- TUNNEL_NO_CHUNKED_ENCODING= #Disables chunked transfer encoding; useful if you are running a WSGI server. (default: false)
     #- TUNNEL_PROXY_ADDRESS= #Listen address for the proxy. (default: "127.0.0.1")
     #- TUNNEL_PROXY_PORT= #Listen port for the proxy. (default: 0)
     #- TUNNEL_LOGLEVEL= #Application logging level {debug, info, warn, error, fatal}. At debug level cloudflared will log request URL, method, protocol, content length, as well as, all request and response headers. This can expose sensitive information in your logs. (default: "info")
     #- TUNNEL_LOGFILE= #Save application log to this file for reporting issues.
     #- TUNNEL_LOGDIRECTORY= #Save application log to this directory for reporting issues.
     #- TUNNEL_TRACE_OUTPUT= #Name of trace output file, generated when cloudflared stops.
     #- TUNNEL_DNS= #Run a DNS over HTTPS proxy server. (default: false)
     #- TUNNEL_DNS_PORT= #Listen on given port for the DNS over HTTPS proxy server. (default: 53)
     #- TUNNEL_DNS_ADDRESS= #Listen address for the DNS over HTTPS proxy server. (default: "localhost")
     #- TUNNEL_DNS_UPSTREAM= #Upstream endpoint URL, you can specify multiple endpoints for redundancy. (default: "https://1.1.1.1/dns-query", "https://1.0.0.1/dns-query")  (accepts multiple inputs)
     #- TUNNEL_DNS_MAX_UPSTREAM_CONNS= #Maximum concurrent connections to upstream. Setting to 0 means unlimited. (default: 5)
     #- TUNNEL_DNS_BOOTSTRAP= #bootstrap endpoint URL, you can specify multiple endpoints for redundancy. (default: "https://162.159.36.1/dns-query", "https://162.159.46.1/dns-query", "https://[2606:4700:4700::1111]/dns-query", "https://[2606:4700:4700::1001]/dns-query")  (accepts multiple inputs)
     #- TUNNEL_HOSTNAME= #Set a hostname on a Cloudflare zone to route traffic through this tunnel.
     #- TUNNEL_LB_POOL= #The name of a (new/existing) load balancing pool to add this origin to.
     #- TUNNEL_TAG= #Custom tags used to identify this tunnel, in format KEY=VALUE. Multiple tags may be specified  (accepts multiple inputs)
     #- TUNNEL_RETRIES= #Maximum number of retries for connection/protocol errors. (default: 5)
     #- TUNNEL_GRACE_PERIOD= #When cloudflared receives SIGINT/SIGTERM it will stop accepting new requests, wait for in-progress requests to terminate, then shutdown. Waiting for in-progress requests will timeout after this grace period, or when a second SIGTERM/SIGINT is received. (default: 30s)
     #- TUNNEL_COMPRESSION_LEVEL= #(beta) Use cross-stream compression instead HTTP compression. 0-off, 1-low, 2-medium, >=3-high. (default: 0) 
     #- TUNNEL_NAME= #Stable name to identify the tunnel. Using this flag will create, route and run a tunnel. For production usage, execute each command separately

version: '3.3'
services:

  #tunnel:
    #container_name: cloudflared-tunnel
    #image: cloudflare/cloudflared:2021.5.6
    #volumes:
    #  - ~/docker/.cloudflared:/home/nonroot/.cloudflared
    #environment:
    #  - TZ=Asia/Kuala_Lumpur
    #command: tunnel route dns <TUNNELNAME> <whatever.example.com>
    #command: tunnel ingress validate
    #command: tunnel --config /home/nonroot/.cloudflared/config.yml --no-autoupdate=false run
    #restart: unless-stopped
  
  tunnel2:
    container_name: cloudflared-tunnel-2
    image: cloudflare/cloudflared:2021.5.8
    volumes:
      - ~/docker/.cloudflared:/home/nonroot/.cloudflared
    environment:
      - TZ=Asia/Kuala_Lumpur
    #command: tunnel route dns <TUNNELNAME> <whatever.example.com>
    #command: tunnel ingress validate
    command: tunnel --config /home/nonroot/.cloudflared/config.yml --no-autoupdate=false run
    restart: unless-stopped

networks:
  default:
    external:
      name: yourdockernetwork
```

**Comment by Kate222** [comment]
I try to run cloudflare tunnel on a rpi 4 with OMV. I have set up tunnel container through portainer.  Could you please share  more details how you do authentication with Cloudflare? @henkisdabro Thank you

**Comment by Kate222** [comment]
Yes, I am able to access OMV through Cloudflare tunnel. I use rpi4.

**Comment by Aniganesh** [comment]
Did you do anything other than what henkisdabro mentioned? If so could you please share? @Kate222

**Comment by Kate222** [comment]
I did it differently. It was a year ago. I will have a look at my notes and
will share it with you if it still works.

**Comment by Kate222** [comment]
I have installed latest rpi os and OMV6 on my rpi4, the cloudflared tunnel is working fine. I am able to access OMV6 from outside. 

This youtube video has given very good information about it. 
 https://www.youtube.com/watch?v=hrwoKO7LMzk&list=PLPUuhipqbOXSr-LIm5PLNojvLw3nIWNOY&index=11

Because OMV is armv71, so the amd64 version in the video does not work with rpi4/omv setup. 
Here are a few things what I have done differently from the video. 

1 install cloudflared
Here we are downloading the precompiled binary and copying it to the /usr/local/bin/ directory to allow execution by the cloudflared user. Proceed to run the binary with the -v flag to check it is all working:

wget https://github.com/cloudflare/cloudflared/releases/latest/download/cloudflared-linux-arm
sudo cp ./cloudflared-linux-arm /usr/local/bin/cloudflared
sudo chmod +x /usr/local/bin/cloudflared
cloudflared -v

2 create a config.yml file

sudo nano /home/yourownusername/.cloudflared/config.yml


tunnel: tunnelidyouhavecreated
credentials-file: /home/yourownusername/.cloudflared/tunneidyouhavecreated.json
ingress:
  - hostname: yourownhostname
    service: http://yourlocalipaddress
    originRequest:
      noTLSVerify: true

  - service: http_status:404

I have found the command line below is useful to check ingress rules before run the tunnel.  
cloudflared tunnel ingress validate

Hope it helps. 
Kate
